### PR TITLE
fix(core): fix OOM on start caused by O3 purge backglog

### PR DESCRIPTION
**Pull request by ideoma**
When `sys.column_versions_purge_log ` table has too many unprocessed records QuestDB can fail to start with Out Of Memory error while trying to re-schedule column purges.

Limit the number of records read from `sys.column_versions_purge_log` and log error suggesting to run `VACUUM TABLE`

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 33 / 33 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|29|29|100.00%|

**Review by glasstiger — DISMISSED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 35 / 43 (81.40%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|2|10|20.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropertyKey.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|29|29|100.00%|

File: core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java [hunk 1]
```diff
@@ -253,6 +268,14 @@ private void putTasksFromTableToQueue() {
                     if (taskRun != null) {
                         retryQueue.add(taskRun);
                     }
+
+                    if (count > columnPurgeQueueLimit) {
+                        // There can be too many tasks in the table so that this code can cause OOM
+                        // and QuestDB will not be able to start at all. Stop here and let the user know
+                        LOG.critical().$("too many Column Purge Tasks in the table some task will be ignored, " +
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java:275]: I am not sure alerting someone at 2am makes any sense when we could have automatically removed unused items on startup, synchronously.

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java (hunk 1 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java]: done

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 63 / 73 (86.30%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|31|41|75.61%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|32|32|100.00%|

File: core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java [hunk 2]
```diff
@@ -251,12 +268,16 @@ private void putTasksFromTableToQueue() {
...
+
+                if (count > 0) {
+                    LOG.info().$("cleaned up rewritten column files [cleanCount=").$(count).I$();
+                }
             }
-            if (retryQueue.size() == 0 && writer != null) {
-                // No tasks to do. Cleanup the log table
+            if (writer != null) {
                 try {
                     writer.truncate();
```
Comment (eugenels) [inline core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java:282]: Looks like incomplete entries older than `PurgeRetryLimitDays` are ignored and possibly lost as the table is unconditionally truncated on every restart.

Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java:282]: I think it's fine. If we couldn't delete file for a month, what are the chances to delete it after that.

File: core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java [hunk 3]
```diff
@@ -203,6 +190,29 @@ private boolean processInQueue() {
...
+                }
+            } else {
+                // All reruns are in the future.
+                return useful;
+            }
+        }
+        return useful;
+    }
+
     private void putTasksFromTableToQueue() {
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java]: renamed to `processTableRecords`

File: core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java (hunk 2 above)
Comment (ideoma) [reply core/src/main/java/io/questdb/cairo/ColumnPurgeJob.java:280]: remove the limit

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 74 (86.49%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|31|41|75.61%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|33|33|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Review by eugenels — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 74 (86.49%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|31|41|75.61%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|33|33|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 64 / 74 (86.49%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeOperator.java|31|41|75.61%|
|:large_blue_circle:|io/questdb/cairo/ColumnPurgeJob.java|33|33|100.00%|

**Pull request merged**